components/driver/sdmmc_transaction.c

```diff
@@ -153,7 +154,11 @@ esp_err_t sdmmc_host_do_transaction(int slot, sdmmc_command_t* cmdinfo)
     cmdinfo->error = ESP_OK;
     sdmmc_req_state_t state = SDMMC_SENDING_CMD;
     sdmmc_event_t unhandled_events = { 0 };
+    int t0 = esp_timer_get_time();
     while (state != SDMMC_IDLE) {
+        if (esp_timer_get_time() - t0 > SDMMC_HOST_DO_TRANSACTION_TIMEOUT_MS) {
```

Comment:
Correct me if I'm wrong, but it is not *just* a state machine. It seems like it does active communication with the SD card, which is why I added a timeout.

For example:

```
static esp_err_t process_events(sdmmc_event_t evt, sdmmc_command_t* cmd,
        sdmmc_req_state_t* pstate, sdmmc_event_t* unhandled_events)

{
    while (next_state != state) {
            state = next_state; // note: this might still be the same state as before!
            switch (state) {

         case SDMMC_SENDING_DATA:
                  .....
                 // it looks like this code will loop until some stop condition is set in the peripheral
                break;
}

```

Reply:
>  If we stay in the same state, then the function exits and we get back to sdmmc_host_wait_for_event

So, could a broken SD card prevent us from exiting? i.e. by returning new events infinitely?

lmk if I should remove this check, or if it is of some value.

Reply:
Not really, it's more like if the SDMMC peripheral suddenly stops following its documented state machine. I mean, it's okay to not trust the hardware, but if we sprinkle all the drivers for checks that the chip peripheral is not following its spec, it will be a mess to read and the code size will be impacted. So I do think this one is unnecessary.

(I agree with the other ones which are card behavior dependent, though.)

Reply:
Okay. removed!